src/processing/UndoRedo.java

```diff
@@ -0,0 +1,281 @@
+	 * @param intensity new intensity or color of the object
+	 * @return newly created color changed object
+	 */
+	private static BoardObject colorChangeOperation(BoardObject obj, Intensity intensity) {
+		UserId id = obj.getUserId();
+		// gets the newly created color changed object
+		BoardObject newObj = ParameterizedOperationsUtil.colorChangeUtil(obj, id, intensity);
+		return newObj;
+	}
+	
+	/**
+	 * This helper method calls rotationUtil method to rotate the object
+	 * It is called during undo-redo operations
+	 * 
+	 * @param obj object that has to be rotated
+	 * @param angle angle in CCW the object has to be rotated
+	 * @return the newly created rotated object
+	 */
+	private static BoardObject rotateOperation(BoardObject obj, Angle angle) {
+		UserId id = obj.getUserId();
+		// gets the newly rotated object
+		BoardObject newObj = ParameterizedOperationsUtil.rotationUtil(obj, id, angle);
+		return newObj;
+	}
+	
+	/**
+	 * This method performs undo or redo operation
+	 * To perform any of the two operation proper arguments need to be passed
+	 * 
+	 * @param curStack could be undo or redo stack depending upon operation
+	 * @param otherStack other stack than curStack
+	 * @param operation undo or redo operation
+	 */
+	private static void undoRedoUtil(
+			ArrayList <BoardObject> curStack,
+			ArrayList <BoardObject> otherStack,
+			Operation operation
+			) {
+		
+		// No undo or redo operation possible
+    	if (curStack.size() <= 0)
+    		return;
+    	
+    	// Gets the top object from the stack
+    	BoardObject topObj =  curStack.get(curStack.size() - 1);
+    	
+    	// Gets the operation performed on that object
+    	BoardObjectOperationType operationType = topObj.getOperation().getOperationType();
+    	BoardObject newObj = null;
+    	
+    	switch (operationType) {
+    		case CREATE :	if (operation == Operation.REDO)
+    							newObj = createOperation(topObj);
+    						else
+    							// performs simple delete operation
+    							ClientBoardState.maps.removeObjectFromMaps(topObj.getObjectId());
+    						break;
+    		case DELETE :	if (operation == Operation.UNDO)
+    							newObj = createOperation(topObj);
+    						else 
+    							// performs simple delete operation
+    							ClientBoardState.maps.removeObjectFromMaps(topObj.getObjectId());
+    						break;
+    					  
+    		case ROTATE : Angle angleCCW = ((RotateOperation) topObj.getOperation()).getAngle();
+    					  Angle newAngle = null;
+    					  /** 
+    					   *  For undo, the object should be rotated -angle CCW
+    					   *  For redo, the object should be rotated angle CCW
+    					   */
+    					  if (operation == Operation.UNDO)
+    						  newAngle = new Angle(-angleCCW.angle);
+    					  else
+    						  newAngle = new Angle(angleCCW.angle);
+    					  newObj = rotateOperation(topObj, newAngle);
+    					  break;
+    					   
+    		case COLOR_CHANGE : ArrayList <Pixel> newPixels;
```

Comment:
Indentation is quite off. Please check if you are using tabs instead of whitespaces.